libs/pdi-scanner/src/rusb_async/pool.rs

```diff
@@ -0,0 +1,269 @@
+pub struct TransferPool<C: UsbContext> {
+    device: Arc<DeviceHandle<C>>,
+    pending: VecDeque<Transfer>,
+}
+
+impl<C: UsbContext> TransferPool<C> {
+    pub fn new(device: Arc<DeviceHandle<C>>) -> Self {
+        Self {
+            device,
+            pending: VecDeque::new(),
+        }
+    }
+
+    #[allow(unused)]
+    pub fn submit_bulk(&mut self, endpoint: u8, buf: Vec<u8>) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::bulk(self.device.as_raw(), endpoint, buf);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    #[allow(unused)]
+    pub fn submit_control(
+        &mut self,
+        request_type: u8,
+        request: u8,
+        value: u16,
+        index: u16,
+        data: &[u8],
+    ) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::control(
+                self.device.as_raw(),
+                request_type,
+                request,
+                value,
+                index,
+                data,
+            );
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    #[allow(unused)]
+    pub unsafe fn submit_control_raw(&mut self, buffer: Vec<u8>) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::control_raw(self.device.as_raw(), buffer);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    #[allow(unused)]
+    pub fn submit_interrupt(&mut self, endpoint: u8, buf: Vec<u8>) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::interrupt(self.device.as_raw(), endpoint, buf);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    #[allow(unused)]
+    pub fn submit_iso(&mut self, endpoint: u8, buf: Vec<u8>, iso_packets: i32) -> Result<()> {
+        // Safety: If transfer is submitted, it is pushed onto `pending` where it will be
+        // dropped before `device` is freed.
+        unsafe {
+            let mut transfer = Transfer::iso(self.device.as_raw(), endpoint, buf, iso_packets);
+            transfer.submit()?;
+            self.pending.push_back(transfer);
+            Ok(())
+        }
+    }
+
+    pub fn poll(&mut self, timeout: Duration) -> Result<CompletedTransfer> {
```

Comment:
This used to just return a `Result<Vec<u8>>`, but that meant I couldn't identify which endpoint the data came from. The `CompletedTransfer` struct gives me both the data and the endpoint.